{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e42314cf-ef21-460f-9bd6-41bd9a663862",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bystro.api import auth\n",
    "from bystro.proteomics.annotation_interface import get_annotation_result_from_query, execute_query, async_get_annotation_result_from_query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca15c9c2-9050-4370-9871-d1c69e82a194",
   "metadata": {},
   "outputs": [],
   "source": [
    "user = auth.login('email', 'password', 'https://bystro-dev.emory.edu')\n",
    "index = '663ac4bfa0e17a1660ba8130_657a50d4b2d0278938ba791d'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e7024b8d-9206-496f-8c8d-71d58a3c208b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>vcfPos</th>\n",
       "      <th>inputRef</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample</th>\n",
       "      <th>dosage</th>\n",
       "      <th>gnomad.exomes.AF</th>\n",
       "      <th>gnomad.genomes.AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2923767</td>\n",
       "      <td>2923767</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2923767:A:G</td>\n",
       "      <td>1805</td>\n",
       "      <td>2</td>\n",
       "      <td>0.918969</td>\n",
       "      <td>0.922388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2923767</td>\n",
       "      <td>2923767</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2923767:A:G</td>\n",
       "      <td>1847</td>\n",
       "      <td>2</td>\n",
       "      <td>0.918969</td>\n",
       "      <td>0.922388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2923767</td>\n",
       "      <td>2923767</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2923767:A:G</td>\n",
       "      <td>4805</td>\n",
       "      <td>2</td>\n",
       "      <td>0.918969</td>\n",
       "      <td>0.922388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929286</td>\n",
       "      <td>2929286</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929286:A:C</td>\n",
       "      <td>1805</td>\n",
       "      <td>2</td>\n",
       "      <td>0.686967</td>\n",
       "      <td>0.695134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929286</td>\n",
       "      <td>2929286</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929286:A:C</td>\n",
       "      <td>1847</td>\n",
       "      <td>2</td>\n",
       "      <td>0.686967</td>\n",
       "      <td>0.695134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929286</td>\n",
       "      <td>2929286</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929286:A:C</td>\n",
       "      <td>4805</td>\n",
       "      <td>2</td>\n",
       "      <td>0.686967</td>\n",
       "      <td>0.695134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929392</td>\n",
       "      <td>2929392</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929392:G:A</td>\n",
       "      <td>1805</td>\n",
       "      <td>1</td>\n",
       "      <td>0.441199</td>\n",
       "      <td>0.393499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929392</td>\n",
       "      <td>2929392</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929392:G:A</td>\n",
       "      <td>1847</td>\n",
       "      <td>1</td>\n",
       "      <td>0.441199</td>\n",
       "      <td>0.393499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929392</td>\n",
       "      <td>2929392</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929392:G:A</td>\n",
       "      <td>4805</td>\n",
       "      <td>2</td>\n",
       "      <td>0.441199</td>\n",
       "      <td>0.393499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr2</td>\n",
       "      <td>196825402</td>\n",
       "      <td>196825402</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr2:196825402:T:C</td>\n",
       "      <td>1847</td>\n",
       "      <td>1</td>\n",
       "      <td>0.131093</td>\n",
       "      <td>0.157932</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom        pos     vcfPos inputRef alt type id               locus  \\\n",
       "0  chr17    2923767    2923767        A   G  SNP  .   chr17:2923767:A:G   \n",
       "1  chr17    2923767    2923767        A   G  SNP  .   chr17:2923767:A:G   \n",
       "2  chr17    2923767    2923767        A   G  SNP  .   chr17:2923767:A:G   \n",
       "3  chr17    2929286    2929286        A   C  SNP  .   chr17:2929286:A:C   \n",
       "4  chr17    2929286    2929286        A   C  SNP  .   chr17:2929286:A:C   \n",
       "5  chr17    2929286    2929286        A   C  SNP  .   chr17:2929286:A:C   \n",
       "6  chr17    2929392    2929392        G   A  SNP  .   chr17:2929392:G:A   \n",
       "7  chr17    2929392    2929392        G   A  SNP  .   chr17:2929392:G:A   \n",
       "8  chr17    2929392    2929392        G   A  SNP  .   chr17:2929392:G:A   \n",
       "9   chr2  196825402  196825402        T   C  SNP  .  chr2:196825402:T:C   \n",
       "\n",
       "  sample  dosage  gnomad.exomes.AF  gnomad.genomes.AF  \n",
       "0   1805       2          0.918969           0.922388  \n",
       "1   1847       2          0.918969           0.922388  \n",
       "2   4805       2          0.918969           0.922388  \n",
       "3   1805       2          0.686967           0.695134  \n",
       "4   1847       2          0.686967           0.695134  \n",
       "5   4805       2          0.686967           0.695134  \n",
       "6   1805       1          0.441199           0.393499  \n",
       "7   1847       1          0.441199           0.393499  \n",
       "8   4805       2          0.441199           0.393499  \n",
       "9   1847       1          0.131093           0.157932  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Default behavior is to show 1 sample per row, e.g. `melt_samples=True`\n",
    "query_result_df_array_of_structs_select_fields = get_annotation_result_from_query(\n",
    "    query_string=\"*\",\n",
    "    index_name=index,\n",
    "    bystro_api_auth=user,\n",
    "    fields=['gnomad.exomes.AF', 'gnomad.genomes.AF']\n",
    ")\n",
    "query_result_df_array_of_structs_select_fields.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b32ce0d2-4acb-48a6-93d5-2bbda65b5445",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>refSeq.name2</th>\n",
       "      <th>refSeq.spID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SENP3</td>\n",
       "      <td>Q9H4L4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SENP3-EIF4A1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SENP3</td>\n",
       "      <td>Q9H4L4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SENP3-EIF4A1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SENP3</td>\n",
       "      <td>Q9H4L4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1893</th>\n",
       "      <td>DTYMK</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1894</th>\n",
       "      <td>DTYMK</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1895</th>\n",
       "      <td>DTYMK</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1896</th>\n",
       "      <td>ING5</td>\n",
       "      <td>[B7Z6R2, Q8WYH8]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1897</th>\n",
       "      <td>ING5</td>\n",
       "      <td>[B7Z6R2, Q8WYH8]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1898 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      refSeq.name2       refSeq.spID\n",
       "0            SENP3            Q9H4L4\n",
       "1     SENP3-EIF4A1              None\n",
       "2            SENP3            Q9H4L4\n",
       "3     SENP3-EIF4A1              None\n",
       "4            SENP3            Q9H4L4\n",
       "...            ...               ...\n",
       "1893         DTYMK              None\n",
       "1894         DTYMK              None\n",
       "1895         DTYMK              None\n",
       "1896          ING5  [B7Z6R2, Q8WYH8]\n",
       "1897          ING5  [B7Z6R2, Q8WYH8]\n",
       "\n",
       "[1898 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# You can choose to also \"melt\" the dataframe by a field,\n",
    "# so that when the field has an array value, it is flattened\n",
    "# When the primary key of the track isn't found, by default we force that \n",
    "# field to have 1 value per row, and warn you of this behavior.\n",
    "# To change it, set `force_flatten_exploded_field=False`\n",
    "query_result_df = get_annotation_result_from_query(\n",
    "    query_string=\"(type:del && (alt:\\-6 || alt:\\-1)) || (cadd:>40)\",\n",
    "    index_name=index,\n",
    "    bystro_api_auth=user,\n",
    "    melt_samples=True,\n",
    "    explode_field='refSeq.name2',\n",
    "    fields=['refSeq.name2', 'refSeq.spID', 'refSeq.name'],\n",
    "    force_flatten_exploded_field=False\n",
    ")\n",
    "query_result_df[['refSeq.name2', 'refSeq.spID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "820b87d3-a16b-4fc2-8769-550b6594655e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>vcfPos</th>\n",
       "      <th>inputRef</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>locus</th>\n",
       "      <th>refSeq.name2</th>\n",
       "      <th>refSeq.name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7470286</td>\n",
       "      <td>7470285</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:7470286:A:-1</td>\n",
       "      <td>SENP3</td>\n",
       "      <td>NM_015670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>7470286</td>\n",
       "      <td>7470285</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:7470286:A:-1</td>\n",
       "      <td>SENP3-EIF4A1</td>\n",
       "      <td>NR_037926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>220427429</td>\n",
       "      <td>220427428</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr2:220427429:A:-1</td>\n",
       "      <td>OBSL1</td>\n",
       "      <td>NM_015311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>220427429</td>\n",
       "      <td>220427428</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr2:220427429:A:-1</td>\n",
       "      <td>OBSL1</td>\n",
       "      <td>NM_001173431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>220427429</td>\n",
       "      <td>220427428</td>\n",
       "      <td>A</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr2:220427429:A:-1</td>\n",
       "      <td>OBSL1</td>\n",
       "      <td>NM_001173408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr2</td>\n",
       "      <td>238835573</td>\n",
       "      <td>238835572</td>\n",
       "      <td>T</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr2:238835573:T:-1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr15</td>\n",
       "      <td>20490550</td>\n",
       "      <td>20490549</td>\n",
       "      <td>T</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:20490550:T:-1</td>\n",
       "      <td>CHEK2P2</td>\n",
       "      <td>NR_038836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr15</td>\n",
       "      <td>23685542</td>\n",
       "      <td>23685541</td>\n",
       "      <td>T</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:23685542:T:-1</td>\n",
       "      <td>GOLGA6L2</td>\n",
       "      <td>NM_001304388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr15</td>\n",
       "      <td>25453230</td>\n",
       "      <td>25453229</td>\n",
       "      <td>G</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:25453230:G:-1</td>\n",
       "      <td>SNHG14</td>\n",
       "      <td>NR_146177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr15</td>\n",
       "      <td>25453230</td>\n",
       "      <td>25453229</td>\n",
       "      <td>G</td>\n",
       "      <td>-1</td>\n",
       "      <td>DEL</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:25453230:G:-1</td>\n",
       "      <td>SNORD115-21</td>\n",
       "      <td>NR_003313</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom        pos     vcfPos inputRef alt type id                locus  \\\n",
       "0  chr17    7470286    7470285        A  -1  DEL  .   chr17:7470286:A:-1   \n",
       "1  chr17    7470286    7470285        A  -1  DEL  .   chr17:7470286:A:-1   \n",
       "2   chr2  220427429  220427428        A  -1  DEL  .  chr2:220427429:A:-1   \n",
       "3   chr2  220427429  220427428        A  -1  DEL  .  chr2:220427429:A:-1   \n",
       "4   chr2  220427429  220427428        A  -1  DEL  .  chr2:220427429:A:-1   \n",
       "5   chr2  238835573  238835572        T  -1  DEL  .  chr2:238835573:T:-1   \n",
       "6  chr15   20490550   20490549        T  -1  DEL  .  chr15:20490550:T:-1   \n",
       "7  chr15   23685542   23685541        T  -1  DEL  .  chr15:23685542:T:-1   \n",
       "8  chr15   25453230   25453229        G  -1  DEL  .  chr15:25453230:G:-1   \n",
       "9  chr15   25453230   25453229        G  -1  DEL  .  chr15:25453230:G:-1   \n",
       "\n",
       "   refSeq.name2   refSeq.name  \n",
       "0         SENP3     NM_015670  \n",
       "1  SENP3-EIF4A1     NR_037926  \n",
       "2         OBSL1     NM_015311  \n",
       "3         OBSL1  NM_001173431  \n",
       "4         OBSL1  NM_001173408  \n",
       "5          None          None  \n",
       "6       CHEK2P2     NR_038836  \n",
       "7      GOLGA6L2  NM_001304388  \n",
       "8        SNHG14     NR_146177  \n",
       "9   SNORD115-21     NR_003313  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# You can disable sample melting by setting `melt_samples=False`\n",
    "query_result_df = await async_get_annotation_result_from_query(\n",
    "    query_string=\"(type:del && (alt:\\-6 || alt:\\-1)) || (cadd:>40)\",\n",
    "    index_name=index,\n",
    "    bystro_api_auth=user,\n",
    "    fields=['refSeq.name2', 'refSeq.name'],\n",
    "    melt_samples=False,\n",
    "    explode_field='refSeq.name2'\n",
    ")\n",
    "query_result_df.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "100870ca-71eb-4af3-a36e-2fbb7f8c9055",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>vcfPos</th>\n",
       "      <th>inputRef</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>id</th>\n",
       "      <th>locus</th>\n",
       "      <th>sample</th>\n",
       "      <th>dosage</th>\n",
       "      <th>...</th>\n",
       "      <th>gnomad.genomes</th>\n",
       "      <th>heterozygosity</th>\n",
       "      <th>homozygosity</th>\n",
       "      <th>missingness</th>\n",
       "      <th>nearest.refSeq</th>\n",
       "      <th>nearestTss.refSeq</th>\n",
       "      <th>ref</th>\n",
       "      <th>refSeq</th>\n",
       "      <th>sampleMaf</th>\n",
       "      <th>trTv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2923767</td>\n",
       "      <td>2923767</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2923767:A:G</td>\n",
       "      <td>1805</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AN_nfe_nwe': 8598, 'AF_fin': 0.9588609933853...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...</td>\n",
       "      <td>[{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...</td>\n",
       "      <td>A</td>\n",
       "      <td>[{'kgID': None, 'spDisplayID': None, 'rfamAcc'...</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2923767</td>\n",
       "      <td>2923767</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2923767:A:G</td>\n",
       "      <td>1847</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AN_nfe_nwe': 8598, 'AF_fin': 0.9588609933853...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...</td>\n",
       "      <td>[{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...</td>\n",
       "      <td>A</td>\n",
       "      <td>[{'kgID': None, 'spDisplayID': None, 'rfamAcc'...</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2923767</td>\n",
       "      <td>2923767</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2923767:A:G</td>\n",
       "      <td>4805</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AN_nfe_nwe': 8598, 'AF_fin': 0.9588609933853...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...</td>\n",
       "      <td>[{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...</td>\n",
       "      <td>A</td>\n",
       "      <td>[{'kgID': None, 'spDisplayID': None, 'rfamAcc'...</td>\n",
       "      <td>1.000</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929286</td>\n",
       "      <td>2929286</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929286:A:C</td>\n",
       "      <td>1805</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AF_nfe_seu': 0.537735998630524, 'AF_afr': 0....</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'NM_001330058', 'dist': 0, 'name2': ...</td>\n",
       "      <td>[{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...</td>\n",
       "      <td>A</td>\n",
       "      <td>[{'siteType': 'intronic', 'altAminoAcid': None...</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929286</td>\n",
       "      <td>2929286</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929286:A:C</td>\n",
       "      <td>1847</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AF_nfe_seu': 0.537735998630524, 'AF_afr': 0....</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'NM_001330058', 'dist': 0, 'name2': ...</td>\n",
       "      <td>[{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...</td>\n",
       "      <td>A</td>\n",
       "      <td>[{'siteType': 'intronic', 'altAminoAcid': None...</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929286</td>\n",
       "      <td>2929286</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929286:A:C</td>\n",
       "      <td>4805</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AF_nfe_seu': 0.537735998630524, 'AF_afr': 0....</td>\n",
       "      <td>0.000</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name': 'NM_001330058', 'dist': 0, 'name2': ...</td>\n",
       "      <td>[{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...</td>\n",
       "      <td>A</td>\n",
       "      <td>[{'siteType': 'intronic', 'altAminoAcid': None...</td>\n",
       "      <td>1.000</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929392</td>\n",
       "      <td>2929392</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929392:G:A</td>\n",
       "      <td>1805</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AF_afr': 0.270516008138657, 'AF_fin': 0.4530...</td>\n",
       "      <td>0.667</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...</td>\n",
       "      <td>[{'name': 'NM_014566', 'dist': 36571, 'name2':...</td>\n",
       "      <td>G</td>\n",
       "      <td>[{'exonicAlleleFunction': 'synonymous', 'altCo...</td>\n",
       "      <td>0.667</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929392</td>\n",
       "      <td>2929392</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929392:G:A</td>\n",
       "      <td>1847</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AF_afr': 0.270516008138657, 'AF_fin': 0.4530...</td>\n",
       "      <td>0.667</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...</td>\n",
       "      <td>[{'name': 'NM_014566', 'dist': 36571, 'name2':...</td>\n",
       "      <td>G</td>\n",
       "      <td>[{'exonicAlleleFunction': 'synonymous', 'altCo...</td>\n",
       "      <td>0.667</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr17</td>\n",
       "      <td>2929392</td>\n",
       "      <td>2929392</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:2929392:G:A</td>\n",
       "      <td>4805</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AF_afr': 0.270516008138657, 'AF_fin': 0.4530...</td>\n",
       "      <td>0.667</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...</td>\n",
       "      <td>[{'name': 'NM_014566', 'dist': 36571, 'name2':...</td>\n",
       "      <td>G</td>\n",
       "      <td>[{'exonicAlleleFunction': 'synonymous', 'altCo...</td>\n",
       "      <td>0.667</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr2</td>\n",
       "      <td>196825402</td>\n",
       "      <td>196825402</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>.</td>\n",
       "      <td>chr2:196825402:T:C</td>\n",
       "      <td>1847</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>{'AN_nfe': 15422, 'AF_nfe': 0.123718999326229,...</td>\n",
       "      <td>0.333</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[{'name2': 'DNAH7', 'name': 'NM_018897', 'dist...</td>\n",
       "      <td>[{'name2': 'STK17B', 'name': 'NM_004226', 'dis...</td>\n",
       "      <td>T</td>\n",
       "      <td>[{'refAminoAcid': 'K', 'kgID': 'uc002utj.4', '...</td>\n",
       "      <td>0.167</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom        pos     vcfPos inputRef alt type id               locus  \\\n",
       "0  chr17    2923767    2923767        A   G  SNP  .   chr17:2923767:A:G   \n",
       "1  chr17    2923767    2923767        A   G  SNP  .   chr17:2923767:A:G   \n",
       "2  chr17    2923767    2923767        A   G  SNP  .   chr17:2923767:A:G   \n",
       "3  chr17    2929286    2929286        A   C  SNP  .   chr17:2929286:A:C   \n",
       "4  chr17    2929286    2929286        A   C  SNP  .   chr17:2929286:A:C   \n",
       "5  chr17    2929286    2929286        A   C  SNP  .   chr17:2929286:A:C   \n",
       "6  chr17    2929392    2929392        G   A  SNP  .   chr17:2929392:G:A   \n",
       "7  chr17    2929392    2929392        G   A  SNP  .   chr17:2929392:G:A   \n",
       "8  chr17    2929392    2929392        G   A  SNP  .   chr17:2929392:G:A   \n",
       "9   chr2  196825402  196825402        T   C  SNP  .  chr2:196825402:T:C   \n",
       "\n",
       "  sample  dosage  ...                                     gnomad.genomes  \\\n",
       "0   1805       2  ...  {'AN_nfe_nwe': 8598, 'AF_fin': 0.9588609933853...   \n",
       "1   1847       2  ...  {'AN_nfe_nwe': 8598, 'AF_fin': 0.9588609933853...   \n",
       "2   4805       2  ...  {'AN_nfe_nwe': 8598, 'AF_fin': 0.9588609933853...   \n",
       "3   1805       2  ...  {'AF_nfe_seu': 0.537735998630524, 'AF_afr': 0....   \n",
       "4   1847       2  ...  {'AF_nfe_seu': 0.537735998630524, 'AF_afr': 0....   \n",
       "5   4805       2  ...  {'AF_nfe_seu': 0.537735998630524, 'AF_afr': 0....   \n",
       "6   1805       1  ...  {'AF_afr': 0.270516008138657, 'AF_fin': 0.4530...   \n",
       "7   1847       1  ...  {'AF_afr': 0.270516008138657, 'AF_fin': 0.4530...   \n",
       "8   4805       2  ...  {'AF_afr': 0.270516008138657, 'AF_fin': 0.4530...   \n",
       "9   1847       1  ...  {'AN_nfe': 15422, 'AF_nfe': 0.123718999326229,...   \n",
       "\n",
       "   heterozygosity  homozygosity missingness  \\\n",
       "0           0.000         1.000         0.0   \n",
       "1           0.000         1.000         0.0   \n",
       "2           0.000         1.000         0.0   \n",
       "3           0.000         1.000         0.0   \n",
       "4           0.000         1.000         0.0   \n",
       "5           0.000         1.000         0.0   \n",
       "6           0.667         0.333         0.0   \n",
       "7           0.667         0.333         0.0   \n",
       "8           0.667         0.333         0.0   \n",
       "9           0.333         0.000         0.0   \n",
       "\n",
       "                                      nearest.refSeq  \\\n",
       "0  [{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...   \n",
       "1  [{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...   \n",
       "2  [{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...   \n",
       "3  [{'name': 'NM_001330058', 'dist': 0, 'name2': ...   \n",
       "4  [{'name': 'NM_001330058', 'dist': 0, 'name2': ...   \n",
       "5  [{'name': 'NM_001330058', 'dist': 0, 'name2': ...   \n",
       "6  [{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...   \n",
       "7  [{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...   \n",
       "8  [{'name2': 'RAP1GAP2', 'name': 'NM_001330058',...   \n",
       "9  [{'name2': 'DNAH7', 'name': 'NM_018897', 'dist...   \n",
       "\n",
       "                                   nearestTss.refSeq ref  \\\n",
       "0  [{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...   A   \n",
       "1  [{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...   A   \n",
       "2  [{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...   A   \n",
       "3  [{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...   A   \n",
       "4  [{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...   A   \n",
       "5  [{'name2': 'OR1D5', 'name': 'NM_014566', 'dist...   A   \n",
       "6  [{'name': 'NM_014566', 'dist': 36571, 'name2':...   G   \n",
       "7  [{'name': 'NM_014566', 'dist': 36571, 'name2':...   G   \n",
       "8  [{'name': 'NM_014566', 'dist': 36571, 'name2':...   G   \n",
       "9  [{'name2': 'STK17B', 'name': 'NM_004226', 'dis...   T   \n",
       "\n",
       "                                              refSeq sampleMaf  trTv  \n",
       "0  [{'kgID': None, 'spDisplayID': None, 'rfamAcc'...     1.000     1  \n",
       "1  [{'kgID': None, 'spDisplayID': None, 'rfamAcc'...     1.000     1  \n",
       "2  [{'kgID': None, 'spDisplayID': None, 'rfamAcc'...     1.000     1  \n",
       "3  [{'siteType': 'intronic', 'altAminoAcid': None...     1.000     2  \n",
       "4  [{'siteType': 'intronic', 'altAminoAcid': None...     1.000     2  \n",
       "5  [{'siteType': 'intronic', 'altAminoAcid': None...     1.000     2  \n",
       "6  [{'exonicAlleleFunction': 'synonymous', 'altCo...     0.667     1  \n",
       "7  [{'exonicAlleleFunction': 'synonymous', 'altCo...     0.667     1  \n",
       "8  [{'exonicAlleleFunction': 'synonymous', 'altCo...     0.667     1  \n",
       "9  [{'refAminoAcid': 'K', 'kgID': 'uc002utj.4', '...     0.167     1  \n",
       "\n",
       "[10 rows x 28 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# You can also represent tracks that have multiple fields by dictionaries, instead of flattening\n",
    "# the dictionaries into separate columns, by setting `structs_of_arrays=False`\n",
    "# This is currently incompatible with `explode_field` and `fields`\n",
    "query_result_df_array_of_structs_no_melt = get_annotation_result_from_query(\n",
    "    query_string=\"*\",\n",
    "    index_name=index,\n",
    "    bystro_api_auth=user,\n",
    "    structs_of_arrays=False\n",
    ")\n",
    "query_result_df_array_of_structs_no_melt.head(n=10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
